core-web/src/main/java/org/silverpeas/core/web/treemenu/process/TreeBuilder.java

```diff
@@ -194,45 +148,172 @@ private static MenuItem buildFirstLevel(TreeFilter filter, String userId, String
     MenuItem item = new MenuItem("root");
     ArrayList<MenuItem> children = new ArrayList<>();
     item.setChildren(children);
-    // the space
-    OrganizationController controller = OrganizationControllerProvider.getOrganisationController();
-    if (filter.acceptNodeType(NodeType.SPACE)) {
-      List<SpaceInstLight> rootSpaces =
-          controller.getRootSpacesContainingComponent(userId, "kmelia");
-      for (SpaceInstLight space : rootSpaces) {
-        if (space != null) {
-          MenuItem subElement =
-              new MenuItem(space.getName(language), space.getId(), 0,
-              NodeType.SPACE, false, null, null);
-          children.add(subElement);
-        }
-      }// the component
-    } else if (filter.acceptNodeType(NodeType.COMPONENT) && !filter.getComponents().isEmpty()) {
+
+    // the spaces
+    MenuItem subElement = getSpaceChildren(userId, language);
+    if (filter.acceptNodeType(NodeType.SPACE) && subElement != null) {
+      children.add(subElement);
+    }
+
```

Comment:
This code doesn't do what the previous one did: browsing the root spaces containing a Kmelia component instance and for each of them creating a MenuItem to add among the children of the root menu item. The code of `getSpaceChildren` is wrong.

Reply:
I refactored as Sonar lint advices to reduce complexity code. It's probably possible to refactor more with more work.